### PR TITLE
QL: Remove implicit conversion inside Literal

### PR DESCRIPTION
**Pull request by costin**
Literal constructor makes an implicit conversion for each value given
which turns out has some subtle side-effects.
Improve MathProcessors to preserve numeric type where possible
Fix bug on issue compatibility between date and intervals
Preserve the source when folding inside the Optimizer

### CODE REVIEW

**Comment by costin** [comment]
@elasticmachine run elasticsearch-ci/2

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java [hunk 1]
```diff
@@ -21,17 +23,38 @@
...
+            }
+
+            // fallback to integer
             long lo = ((Number) l).longValue();
-            //handles the corner-case of Long.MIN_VALUE
-            return lo >= 0 ? lo : lo == Long.MIN_VALUE ? Double.valueOf(Long.MAX_VALUE) : -lo;
-        }),
 
+            if (lo == Long.MIN_VALUE) {
+                throw new QlIllegalArgumentException("[" + lo + "] cannot be negated since the result is outside the range");
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java]: Is this case tested?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java [hunk 2]
```diff
@@ -21,17 +23,38 @@
...
+            }
+
+            lo = lo < 0 ? -lo : lo;
+
+            if (l instanceof Integer) {
+                if (lo == Integer.MIN_VALUE) {
+                    throw new QlIllegalArgumentException("[" + lo + "] cannot be negated since the result is outside the range");
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java]: This also?

**Review by matriv — APPROVED** [review]
LGTM. Left 2 comments regarding testing.

Moreover, shouldn't we backport the fix regarding the DATE +/- time INTERVAL?

**Comment by costin** [comment]
> Moreover, shouldn't we backport the fix regarding the DATE +/- time INTERVAL?

It will once the 7.x freeze passes and the QL extraction stabilizes.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java]: It is now

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/MathProcessor.java]: Added basic test

**Review by bpintea — APPROVED** [review]

**Comment by matriv** [comment]
> > Moreover, shouldn't we backport the fix regarding the DATE +/- time INTERVAL?
> 
> It will once the 7.x freeze passes and the QL extraction stabilizes.

I ment to previous versions, 7.5 & 6.8

**Pull request merged**